service/role.go

```diff
@@ -145,109 +215,174 @@ func (r *roleService) getRoleToken(ctx context.Context, domain, role, proxyForPr
-		u := getRoleTokenAthenzURL(r.athenzURL, domain, role, minExpiry, maxExpiry, proxyForPrincipal)
+	return echan
+}
 
-		req, err := http.NewRequest(http.MethodGet, u, nil)
-		if err != nil {
-			return nil, err
-		}
+// handleExpiredHook is a handler function for gache expired hook
+func (r *roleService) handleExpiredHook(ctx context.Context, key string) {
+	glg.Debugf("handleExpiredHook(), key: %s", key)
+	domain, role, principal := decode(key)
+	glg.Debugf("decoded: %s, %s, %s", domain, role, principal)
+	_, err := r.updateRoleToken(ctx, domain, role, principal, r.expiry, r.expiry)
+	glg.Errorf("err: %v", err)
+}
 
-		req.Header.Set(r.athenzPrincipleHeader, tok)
+func (r *roleService) updateRoleTokenWithRetry(ctx context.Context, domain, role, proxyForPrincipal string, minExpiry, maxExpiry time.Duration) <-chan error {
+	glg.Debugf("updateRoleTokenWithRetry started, domain: %s, role: %s, proxyForPrincipal: %s, minExpiry: %s, maxExpiry: %s", domain, role, proxyForPrincipal, minExpiry, maxExpiry)
 
-		// send http request
-		res, err := r.httpClient.Do(req.WithContext(ctx))
+	echan := make(chan error, r.errRetryMaxCount)
+	go func() {
+		defer close(echan)
 
-		if err != nil {
-			return nil, err
-		}
-		defer func() {
-			if res != nil {
-				io.Copy(ioutil.Discard, res.Body)
-				if res.Body != nil {
-					res.Body.Close()
-				}
+		for i := 0; i < r.errRetryMaxCount; i++ {
+			if _, err := r.updateRoleToken(ctx, domain, role, proxyForPrincipal, minExpiry, maxExpiry); err != nil {
+				echan <- err
+				time.Sleep(r.errRetryInterval)
+			} else {
+				break
 			}
-		}()
-
-		if res.StatusCode != http.StatusOK {
-			return nil, ErrRoleTokenRequestFailed
 		}
+	}()
 
-		// decode the token as a role token
-		var data *RoleToken
-		err = json.NewDecoder(res.Body).Decode(&data)
-		if err != nil {
-			return nil, err
+	return echan
+}
+
+// updateRoleToken returns RoleToken struct or error.
+// This function ask athenz to generate role token and return, or return any error when generating the role token.
+func (r *roleService) updateRoleToken(ctx context.Context, domain, role, proxyForPrincipal string, minExpiry, maxExpiry time.Duration) (*RoleToken, error) {
+	key := encode(domain, role, proxyForPrincipal)
+
+	rt, err, _ := r.group.Do(key, func() (interface{}, error) {
+		rt, e := r.fetchRoleToken(ctx, domain, role, proxyForPrincipal, minExpiry, maxExpiry)
+		if e != nil {
+			return nil, e
 		}
 
-		// set the token into cache
-		r.domainRoleCache.SetWithExpire(encode(domain, role), &cacheData{
-			token:             data,
+		r.domainRoleCache.SetWithExpire(key, &cacheData{
+			token:             rt,
 			domain:            domain,
 			role:              role,
 			proxyForPrincipal: proxyForPrincipal,
 			minExpiry:         minExpiry,
 			maxExpiry:         maxExpiry,
-		}, time.Unix(data.ExpiryTime, 0).Sub(fastime.Now())-time.Minute)
-		return data, nil
+		}, time.Unix(rt.ExpiryTime, 0).Sub(fastime.Now()))
```

Comment:
better keep the margin and create a margin constant?

Reply:
I think we don't need margin as we update it periodically and if we set margin the cache will be deleted before it is expired.

Reply:
authorizer has it. better to be consistent?
https://github.com/yahoojapan/athenz-authorizer/blob/master/policy/fetcher.go#L48

just hard code to 0 is also fine

Reply:
I think it will make the cache will expire earlier than it should

Reply:
But, when user get a token very close to expire, and send it to other server.
The other side may consider it as an expired token, and reject the request.

need some margin buffer for the process time and network traffic

Reply:
@kevindiu ping

Reply:
ok~